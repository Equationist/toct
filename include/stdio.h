/* stdio.h - Standard Input/Output Library for CompilerKit C Compiler */

#ifndef _STDIO_H
#define _STDIO_H

/* Standard I/O types */
typedef struct {
    int _fd;
    int _flags;
    char* _buffer;
} FILE;

/* Standard streams */
extern FILE* stdin;
extern FILE* stdout; 
extern FILE* stderr;

/* Standard I/O functions */
int printf(const char* format, ...);
int fprintf(FILE* stream, const char* format, ...);
int sprintf(char* str, const char* format, ...);
int snprintf(char* str, int size, const char* format, ...);

int scanf(const char* format, ...);
int fscanf(FILE* stream, const char* format, ...);
int sscanf(const char* str, const char* format, ...);

int putchar(int c);
int puts(const char* s);
int fputs(const char* s, FILE* stream);
int fputc(int c, FILE* stream);

int getchar(void);
char* gets(char* s);
char* fgets(char* s, int size, FILE* stream);
int fgetc(FILE* stream);
int ungetc(int c, FILE* stream);

FILE* fopen(const char* filename, const char* mode);
FILE* freopen(const char* filename, const char* mode, FILE* stream);
int fclose(FILE* stream);
int fflush(FILE* stream);

long ftell(FILE* stream);
int fseek(FILE* stream, long offset, int whence);
void rewind(FILE* stream);
int feof(FILE* stream);
int ferror(FILE* stream);
void clearerr(FILE* stream);

/* File position constants */
#define SEEK_SET 0
#define SEEK_CUR 1
#define SEEK_END 2

/* End of file constant */
#define EOF (-1)

/* Buffer size */
#define BUFSIZ 8192

/* Null pointer */
#ifndef NULL
#define NULL ((void*)0)
#endif

#endif /* _STDIO_H */