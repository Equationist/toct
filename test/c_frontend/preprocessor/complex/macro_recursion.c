/* Test macro recursion and self-reference */
#define EMPTY()
#define DEFER(id) id EMPTY()
#define OBSTRUCT(...) __VA_ARGS__ DEFER(EMPTY)()

/* Self-referential macro (should not expand infinitely) */
#define FOO FOO
#define BAR(x) BAR(x)

/* Recursive-like macro using deferred expansion */
#define REPEAT_1(m, x) m(x)
#define REPEAT_2(m, x) m(x) REPEAT_1(m, x)
#define REPEAT_3(m, x) m(x) REPEAT_2(m, x)

#define PRINT_NUM(x) printf("%d ", x);

#include <stdio.h>

int main() {
    printf("Macro recursion tests:\n");
    
    /* These should just expand to themselves, not infinitely */
    printf("Self-referential macros (should not cause infinite expansion):\n");
    printf("FOO expands to: FOO (literal)\n");
    
    printf("\nRepeat patterns:\n");
    printf("REPEAT_1: ");
    REPEAT_1(PRINT_NUM, 1);
    printf("\n");
    
    printf("REPEAT_2: ");
    REPEAT_2(PRINT_NUM, 2);
    printf("\n");
    
    printf("REPEAT_3: ");
    REPEAT_3(PRINT_NUM, 3);
    printf("\n");
    
    return 0;
}