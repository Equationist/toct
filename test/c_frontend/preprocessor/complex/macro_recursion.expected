Macro recursion tests:
Self-referential macros (should not cause infinite expansion):
FOO expands to: FOO (literal)

Repeat patterns:
REPEAT_1: 1 
REPEAT_2: 2 2 
REPEAT_3: 3 3 3